kong/tools/timestamp.lua

```diff
@@ -5,56 +5,71 @@
 -- @module kong.tools.timestamp
 
 local luatz = require "luatz"
-local _M = {}
+local tz_time = luatz.time
+local tt_from_timestamp = luatz.timetable.new_from_timestamp
+local tt = luatz.timetable.new
+local math_floor = math.floor
 
 --- Current UTC time
--- @return UTC time
-function _M.get_utc()
-  return math.floor(luatz.time()) * 1000
+-- @return UTC time in milliseconds since epoch, but with SECOND precision.
+local function get_utc()
+  return math_floor(tz_time()) * 1000
 end
 
-function _M.get_timetable(now)
-  local timestamp = now and now or _M.get_utc()
-  if #tostring(timestamp) == 13 then
-    timestamp = timestamp / 1000
-  end
-  return luatz.timetable.new_from_timestamp(timestamp)
+--- Current UTC time
+-- @return UTC time in milliseconds since epoch.
+local function get_utc_ms()
+  return tz_time() * 1000
 end
 
---- Creates a timestamp
--- @param now (optional) Time to generate a timestamp from, if omitted current UTC time will be used
--- @return Timestamp table containing fields; second, minute, hour, day, month, year
-function _M.get_timestamps(now)
-  local timetable = _M.get_timetable(now)
+-- setup a validation value, any value above this is assumed to be in MS 
+-- instead of S (a year value beyond the year 20000), it assumes current times
+-- as in 2016 and later.
+local ms_check = tt(20000 , 1 , 1 , 0 , 0 , 0):timestamp()
 
-  local second = luatz.timetable.new(timetable.year, timetable.month,
-                                     timetable.day, timetable.hour,
-                                     timetable.min, timetable.sec)
+-- Returns a time-table.
+-- @param now (optional) time to generate the time-table from. If omitted
+-- current utc will be used. It can be specified either in seconds or 
+-- milliseconds, it will be converted automatically.
+local function get_timetable(now)
+  local timestamp = now and now or get_utc()
+  if timestamp > ms_check then
+    return tt_from_timestamp(timestamp/1000)
+  end
+  return tt_from_timestamp(timestamp)
+end
 
-  local minute = luatz.timetable.new(timetable.year, timetable.month,
-                                     timetable.day, timetable.hour,
-                                     timetable.min, 0)
+--- Creates a timestamp table containing time by different precision levels.
+-- @param now (optional) Time to generate timestamps from, if omitted current UTC time will be used
+-- @return Timestamp table containing fields/precisions; second, minute, hour, day, month, year
+local function get_timestamps(now)
+  local timetable = get_timetable(now)
+  local stamps = {}
```

Comment:
`table.new(0, 6)` would be quite beneficial here since we already know the number of records it will hold in its hash part. Just throwing this as a suggestion for future reference.